pkg/grpc/databroker/sync_cache.go

```diff
@@ -0,0 +1,293 @@
+	}
+
+	return nil
+}
+
+func (c *syncCache) Records(recordType string) iter.Seq2[*Record, error] {
+	prefix := c.recordPrefix(recordType)
+	iterOptions := new(pebble.IterOptions)
+	if c.iterOptions != nil {
+		*iterOptions = *c.iterOptions
+	}
+	iterOptions.LowerBound = prefix
+	iterOptions.UpperBound = pebbleutil.PrefixToUpperBound(prefix)
+	return func(yield func(*Record, error) bool) {
+		for record, err := range pebbleutil.Iterate(c.db, iterOptions, pebbleIteratorToRecord) {
+			if err != nil {
+				yield(nil, fmt.Errorf("sync-cache: error iterating over cached records (record-type=%s): %w", recordType, err))
+				return
+			}
+
+			if !yield(record, nil) {
+				return
+			}
+		}
+	}
+}
+
+func (c *syncCache) Sync(ctx context.Context, client DataBrokerServiceClient, recordType string) error {
+	serverVersion, recordVersion := wrapperspb.UInt64(0), wrapperspb.UInt64(0)
+	err := errors.Join(
+		c.pebbleGetProto(c.db, c.serverVersionKey(recordType), serverVersion),
+		c.pebbleGetProto(c.db, c.recordVersionKey(recordType), recordVersion),
+	)
+	if errors.Is(err, pebble.ErrNotFound) {
+		// if we've never synced, use sync latest
+		return c.syncLatest(ctx, client, recordType)
+	} else if err != nil {
+		return fmt.Errorf("sync-cache: error retrieving server and record version from cache (record-type=%s): %w", recordType, err)
+	}
+
+	return c.sync(ctx, client, recordType, serverVersion.Value, recordVersion.Value)
+}
+
+func (c *syncCache) recordKey(recordType, recordID string) []byte {
+	return slices.Concat(c.recordPrefix(recordType), []byte(recordID))
+}
+
+func (c *syncCache) recordPrefix(recordType string) []byte {
+	return append(c.recordTypePrefix(recordType), fieldRecord)
+}
+
+func (c *syncCache) recordTypePrefix(recordType string) []byte {
+	id := uuid.NewSHA1(syncCacheUUIDNamespace, []byte(recordType))
+	return slices.Concat(c.prefix, id[:])
+}
+
+func (c *syncCache) recordVersionKey(recordType string) []byte {
+	return append(c.recordTypePrefix(recordType), fieldRecordVersion)
+}
+
+func (c *syncCache) serverVersionKey(recordType string) []byte {
+	return append(c.recordTypePrefix(recordType), fieldServerVersion)
+}
+
+func (c *syncCache) sync(ctx context.Context, client DataBrokerServiceClient, recordType string, serverVersion, recordVersion uint64) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	stream, err := client.Sync(ctx, &SyncRequest{
+		Type:          recordType,
+		ServerVersion: serverVersion,
+		RecordVersion: recordVersion,
+		Wait:          proto.Bool(false),
+	})
+	if err != nil {
+		return fmt.Errorf("sync-cache: error starting sync stream (record-type=%s): %w", recordType, err)
+	}
+
+	// batch the updates together
+	batch := c.db.NewBatch()
+	defer batch.Close()
+
+	for {
+		res, err := stream.Recv()
+		if errors.Is(err, io.EOF) {
+			break
+		} else if status.Code(err) == codes.Aborted {
+			// the server version changed, so use sync latest
+			return c.syncLatest(ctx, client, recordType)
```

Comment:
do we need to abort the existing stream here ?

Reply:
oh, but not until sync latest completes... let me refactor that.